---
title: "Preliminary Annotation"
output: html_document
date: '2023-05-22'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Intro and setup

Goal here is to validate the initial clustering and UMAP and ensure things like batch effect and clustering reflect the underlying biology. 


```{r Envrionment}
#+ message = FALSE, warning = FALSE
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(EnsDb.Hsapiens.v86))
suppressPackageStartupMessages(library(BSgenome.Hsapiens.UCSC.hg38))
suppressPackageStartupMessages(library(Matrix))
suppressPackageStartupMessages(library(SeuratWrappers))
suppressPackageStartupMessages(library(harmony))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(dplyr))

setwd("~/gibbs/DOGMAMORPH/Ranalysis")

results<-readRDS( "Objects/20230518Completeobj.rds")

```

## Initial checks

First I am going to check there is no library or individual driven batch effect across the UMAP by visualizing on those. I'll also visualize the treatment and timepoint to see if there are differences there that we're picking up. 

```{r Initial plots, fig.width= 16, fig.height=9}
DimPlot(results, label = TRUE, reduction = "umap")
DimPlot(results, group.by="orig.ident", reduction = "umap")
DimPlot(results, group.by = "Participant", reduction = "umap")
DimPlot(results, group.by = "Treatment", reduction = "umap")
DimPlot(results, group.by = "Timepoint", reduction = "umap")
DimPlot(results, group.by = "Cell.Source", reduction = "umap")
```

Notes:
Clustering - Overall looks good, appears that the substructure of outlying populations is reasonably separated by the clustering algorithm, although the mass of presumably CD4+ T cells are potentially overclustered as we can see several overlapping and presumably small clusters in that mass. 
Library - It appears there are no clusters driven entirely by library. There are differences in distribution but I suspect those are primarily due to PBMC vs CD4 libraries. 
Participant - There is clearly some variation, in participant that is not entirely corrected for (ex there is a patch of green that may be 5018 or something else along the bottom left of the large cluster). Overall this may be acceptbale variation, but after annotation I should check that there are not too much participant - to - participant variability in cluster membership 
Treatment - Minor treatment driven differences but overall appears well mixed 
Time point - Appears to be some outlying populations that are potentially perturbed, but it's overall looking well mixed. 
Cell source - looks like that main cluster is CD4, with a sub cluster pulled out to the left and to the bottom. CD4 also participates in other clusters, likely due to low levels of contamination. 


Also checking here to see if there are any QC related factors that might be driving some of the cluster distribution:

```{r QC plots, fig.width= 16, fig.height=9}
FeaturePlot(results, "mt", reduction = "umap")
FeaturePlot(results, "nCount_RNA", reduction = "umap")
FeaturePlot(results,"nFeature_RNA", reduction = "umap")
FeaturePlot(results,"nCount_ATAC", reduction = "umap")
FeaturePlot(results,"nFeature_ATAC", reduction = "umap")
FeaturePlot(results,"nFeature_ADT.dsb", reduction = "umap")
FeaturePlot(results,"nFeature_ADT.dsb", reduction = "umap")
```

No clusters appear overwhelmingly driven by a QC metric. A few on the periphery have high RNA/ATAC counts/features but nothing too abnormal

Next I'll just briefly look to see if there are quality differences across samples 

```{r Checking QC metrics, fig.width= 16, fig.height=9}
group_by(results@meta.data, Participant)%>%summarise(Cell_count=n(), Median_genes=median(nFeature_RNA), Median_RNA_UMI=median(nCount_RNA), Median_ATAC_Fragments=median(nCount_ATAC), Median_ADT_UMI=median(nCount_ADT))

group_by(results@meta.data, Participant, Timepoint)%>%summarise(Cell_count=n(), Median_genes=median(nFeature_RNA), Median_RNA_UMI=median(nCount_RNA), Median_ATAC_Fragments=median(nCount_ATAC), Median_ADT_UMI=median(nCount_ADT))%>%print(n=10000)

group_by(results@meta.data, Treatment, Timepoint)%>%summarise(Cell_count=n(), Median_genes=median(nFeature_RNA), Median_RNA_UMI=median(nCount_RNA), Median_ATAC_Fragments=median(nCount_ATAC), Median_ADT_UMI=median(nCount_ADT))

```

## Gene Expression overlays 

Goal of this section is to use a few known markers to very roughly annotate the cells. I'll use several sources for these. One is the Seurat tutorial, which is mainly RNA markers: 
Cluster ID	Markers	Cell Type
IL7R, CCR7	Naive CD4+ T
CD14, LYZ	CD14+ Mono
IL7R, S100A4	Memory CD4+
MS4A1	B
CD8A	CD8+ T
FCGR3A, MS4A7	FCGR3A+ Mono
GNLY, NKG7	NK
FCER1A, CST3	DC
PPBP	Platelet



```{r PBMC markers, fig.width= 16, fig.height=9}
pbmc_marks_Seurat<-c("IL7R","CCR7","CD14","LYZ","S100A4","MS4A1","CD8A","FCGR3A","MS4A7","GNLY","NKG7","FCER1A","CST3","PPBP")
DefaultAssay(results)<-"RNA"
results<-ScaleData(results, features = pbmc_marks_Seurat)
FeaturePlot(results, pbmc_marks_Seurat, ncol = 4)
DotPlot(results, features = pbmc_marks_Seurat)+theme(axis.text.x = element_text(vjust = 0.7, angle = 90))
VlnPlot(results, features = pbmc_marks_Seurat)+theme(axis.text.x = element_text(vjust = 0.7, angle = 90))
DoHeatmap(results, pbmc_marks_Seurat)
results<-ScaleData(results, features = pbmc_marks_Seurat, split.by = "orig.ident")
DoHeatmap(results, pbmc_marks_Seurat, disp.min = -2, disp.max = 2)

```

Observations - The heatmap looks somewhat different from what I'm used to on account of it being a mixed population without expression of most markers in most cell types (as opposed to the typical CD4 or the DE defined markers). 
Based on this, I would annotate the first 15 clusters as :
0- Naive CD4+
1- Naive CD4+
2- Mix of Naive and Memory CD4+
3- Memory CD4+
4- Memory CD4+/CD8/NK
5- B cells
6- Naive CD4
7-Memory CD4+
8-Naive CD4+
9-NK
10-B cells
11-CD14+ Mono
12-DC
13-FCGR3A+ Mono
14- Memory CD4+

(no platelets in our data)


Another is the clonal replacement paper for Howard Chang (Yost et al. 2019), also RNA: 
Relevant method Excerpt:
>including CD3G, CD3D, CD3E, CD2 (T cells), CD8A, GZMA (CD8+ T cells), CD4, FOXP3 (CD4+ T cells and Treg cells), KLRC1, KLRC3 (NK cells), CD19, CD79A (B cells), SLAMF7, IGKC (plasma cells), FCGR2A, CSF1R (macrophages), FLT3 (dendritic cells), CLEC4C (plasmacytoid dendritic cells) 

```{r YOST markers, fig.width= 16, fig.height=9}
pbmc_marks_Chang<-c("CD3G", "CD3D", "CD3E", "CD2","CD8A", "GZMA","CD4", "FOXP3","KLRC1", "KLRC3","CD19", "CD79A","SLAMF7", "IGKC" ,"FCGR2A", "CSF1R","FLT3","CLEC4C")

DefaultAssay(results)<-"RNA"
results<-ScaleData(results, features = pbmc_marks_Chang)
FeaturePlot(results, pbmc_marks_Chang, ncol = 4)
DotPlot(results, features = pbmc_marks_Chang)+theme(axis.text.x = element_text(vjust = 0.7, angle = 90))
VlnPlot(results, features = pbmc_marks_Chang)+theme(axis.text.x = element_text(vjust = 0.7, angle = 90))
DoHeatmap(results, pbmc_marks_Chang)
results<-ScaleData(results, features = pbmc_marks_Chang, split.by = "orig.ident")
DoHeatmap(results, pbmc_marks_Chang, disp.min = -2, disp.max = 2)

```

Observations - The heatmap looks somewhat different from what I'm used to on account of it being a mixed population without expression of most markers in most cell types (as opposed to the typical CD4 or the DE defined markers). 
Based on this, I would annotate the first 15 clusters as :
0- T cells
1- T cells
2- T cells
3- T cells
4- T cells/CD8
5- B cells/Plasma
6- T cells
7- T cells/TREG
8- T cells
9- NK
10- B cells/Plasma
11-macrophages/DCs
12-macrophages
13-macrophages
14- ? 

* SLAMF7 and IGKC were discordant. Annotating plasma cells based on just IGKC, appears 15 might be the true plasma cells
* No cluster marked for CLEC4C, pDCs are rare so it might make sense we dont pick them up as a separate cluster, possibly they are a part of cluster 12

Finally we have those that I picked for CD4 more specifically in the immunity paper (Collora et al. 2022), RNA and protein:

Cluster 	Transcriptomic marker 	Surface protein marker positive
GZMB Th1	GZMB, CCL5, TBX21	CD45RO, HLA DR
GZMK Th1	TBX21, GZMK	CD45RO, IL7R, CXCR3, CCR6, CD69
Th2	GATA3	CD45RO, IL7R, CCR4, CD69
Th17	RORC, CTSH, CCR6	CD45RO, IL7R, CCR6, TIGIT
Treg	FOXP3, IL2RA	CD45RO, TIGIT, CD25, HLA DR
Proliferating	MKI67	CD45RO, IL7R, CXCR3, CCR6, CD40L, CD137, PD1, TIGIT, HLA-DR
CXCR5 Memory	CXCR5	CD45RO
Memory		CD45RO, PD1
Naïve	CCR7, SELL, TCF7	CD45RA, IL7R
MT	Percent mitochondria genes	CD45RO


```{r Collora markers, fig.width= 16, fig.height=9}
CD4_marks_Collora_RNA<-c("GZMB", "CCL5", "TBX21", "GZMK","GATA3","RORC", "CTSH","FOXP3", "IL2RA","MKI67", "CXCR5", "CCR7", "SELL", "TCF7", "mt")

DefaultAssay(results)<-"RNA"
results<-ScaleData(results, features = CD4_marks_Collora_RNA)
FeaturePlot(results, CD4_marks_Collora_RNA, ncol = 4)
DotPlot(results, features = CD4_marks_Collora_RNA) +theme(axis.text.x = element_text(vjust = 0.7, angle = 90))
VlnPlot(results, features = CD4_marks_Collora_RNA) +theme(axis.text.x = element_text(vjust = 0.7, angle = 90))
DoHeatmap(results, CD4_marks_Collora_RNA[1:14])
results<-ScaleData(results, features = CD4_marks_Collora_RNA, split.by = "orig.ident")
DoHeatmap(results, CD4_marks_Collora_RNA[1:14], disp.min = -2, disp.max = 2)

CD4_marks_Collora_protein<-c("CD45RA-TotalA", "CD45RO-TotalA", "HLA-DR-DP-DQ-TotalA", "CD127-TotalA","CD183-TotalA","CD196-TotalA","CD194-TotalA","TIGIT-TotalA","CD25-TotalA", "CD154-TotalA", "CD137-TotalA", "CD279-TotalA")
DefaultAssay(results)<-"ADT.dsb"
results<-ScaleData(results, features = CD4_marks_Collora_protein)
FeaturePlot(results, CD4_marks_Collora_protein, ncol = 4, reduction = "umap", min.cutoff = 'q5', max.cutoff = 'q95')
DotPlot(results, features = CD4_marks_Collora_protein, scale.min=0, scale.max=100)+theme(axis.text.x = element_text(vjust = 0.7, angle = 90))
VlnPlot(results, features = CD4_marks_Collora_protein)+theme(axis.text.x = element_text(vjust = 0.7, angle = 90))
DoHeatmap(results, CD4_marks_Collora_protein)

```
Observations and rough annotation, note these markers are for CD4 T cells so we're probably not going to be able to annotate clusters other than 0-8 (minus 5)
RNA alone- 
0- Naive/Memory CD4
1- Naive/Memory CD4
2- Naive/Memory CD4
3- Naive/Memory CD4
4- GZMB+ TH1
5- 
6- Naive CD4
7- Treg, proliferating
8- Memory CD4
9- GZMB+ TH1
10- 
11-
12-
13-
14- 
Protein alone- 
0- RA positive
1- RO positive IL7R positive CXCR3 Positive CCR6 Positive CCR4 Positive TIGIT Positive CD25 positive CD154 Positive CD137 Positive CD279 Positive 
2- RO positive IL7R positive CXCR3 Positive CD154 Positive CD137 Positive
3- RO positive IL7R positive CXCR3 Positive 
4- RO positive
5- DR Positive,  RA positive
6- RA positive
7- DR Positive RO positive, CD25 positive
8- CD137 Positive
9- RA positive
10- DR Positive
11-
12- DR Positive
13-
14- 

Composite annotation (Collora markers)-
0- Naive CD4 T cells 
1- Memory CD4, effector polarized
2- Memory CD4, TFh polarized?
3- Memory CD4
4- Memory CD4 
5- ?
6- Naive CD4
7- Treg, Activated cells 
8- ?
9- Effector/GZMB Th1 
10- 
11-
12- 
13-
14- 

## Final initial clustering across markers for the first 15 clusters

This is the annotation I came up with for the first 15 clusters based on the markers above. 

0- Naive CD4 T cells 
1- Memory CD4, effector polarized
2- Memory CD4, TFh polarized?
3- Memory CD4
4- Memory CD4 Th1 GZMB+
5- B cells
6- Naive CD4
7- Treg, Activated cells 
8- Naive T cells
9- Effector/GZMB Th1/ NK
10-B cells
11-CD14+ Mono
12-DC
13-FCGR3A+ Mono
14-Memory CD4+

Some initial observations: 
The remaining clusters are small and likely cna be merged into other clusters based on some criteria. 
7 Possibly is comprized of T regs and proliferating cells. 
4/9 is the cytotoxic section 
We're currently under utilizing the protein markers for annotation, but further annotation will help with that. 
1 and 2 appear to be merging across several clusters including subsets that are PD1 positive, that have different effector profiles, etc. We will likely need to do a separate clustering focused on specific subsets to parse the heterogeneity here. 

```{r session info}
devtools::session_info()
```


