---
title: "Proliferating mod ID version 2"
author: "Jack Collora"
date: '2023-06-01'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r setup functions,echo=FALSE}
## Required packages for code
suppressPackageStartupMessages(require(WGCNA))
suppressPackageStartupMessages(require(flashClust))
suppressPackageStartupMessages(require(Hmisc))
suppressPackageStartupMessages(require(dplyr))
suppressPackageStartupMessages(library(doParallel))
suppressPackageStartupMessages(require(sva))

registerDoParallel(cores=10)
options(warn=0)

## Required packages for saving results
suppressPackageStartupMessages(require(openxlsx))
suppressPackageStartupMessages(require(ggplot2))
suppressPackageStartupMessages(require(cowplot))
suppressPackageStartupMessages(library(Seurat))
#these functions are from Sam Kazar's 2020 Nature Med paper with minor modifications and added notes
{# Choosing the appropriate power for generating the adjacency matrix.
  FindPower <- function(datExpr){
    #choose soft-threshold power
    powers = c(c(1:10), seq(from = 12, to=20, by=2))
    sft=pickSoftThreshold(datExpr,dataIsExpr = TRUE,powerVector = powers,corFnc = cor,
                          corOptions = list(use = 'p', method = "pearson"),networkType = "signed")
    
    # Plot the results
    par(mfrow = c(1,2));
    cex1 = 0.9;
    
    # Scale-free topology fit index as a function of the soft-thresholding power
    plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit, signed R^2",type="n", main = paste("Scale independence"));
    text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],labels=powers,cex=cex1,col="red");
    
    # Red line corresponds to using an R^2 cut-off
    abline(h=0.80,col="red")
    
    # Mean connectivity as a function of the soft-thresholding power
    plot(sft$fitIndices[,1], sft$fitIndices[,5],xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",main = paste("Mean connectivity"))
    text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
  }
  
  # Generating the adjacency matrix and performing clustering
  ClusterTOM <- function(datExpr, softPower){
    #dev.off()
    #Calclute the adjacency matrix
    adj= adjacency(datExpr,type = "signed", power = softPower);
    
    #Turn adjacency matrix into a topological overlap matrix (TOM) to minimize the effects of noise and spurious associations.
    TOM=TOMsimilarityFromExpr(datExpr,networkType = "signed", TOMType = "signed", power = softPower, corType="bicor");
    
    colnames(TOM) = rownames(TOM) = colnames(datExpr)
    dissTOM=1-TOM
    
    #Hierarchical clustering of the genes based on the TOM dissimilarity measure
    geneTree = flashClust(as.dist(dissTOM),method="complete");
    
    #Plot the resulting clustering tree (dendrogram)
    plot(geneTree, xlab="", sub="",cex=0.3);
    
    return(list(dissTOM = dissTOM, geneTree = geneTree)) #returns list with dissimilarity TOM, and the clustered gene tree.
  }
  
  # Cut the resulting clustering dendrogram using the "tree" method for cutreeDynamic. Minimum module size can be specified.
  CutTOMTree <- function(datExpr, dissTOM, geneTree, minModuleSize = 10){
    #dev.off()
    # Module identification using dynamic tree cut, you can also choose the hybrid method
    dynamicMods = cutreeDynamic(dendro = geneTree,  method="tree", minClusterSize = minModuleSize);
    #dynamicMods = cutreeDynamic(dendro = geneTree, distM = dissTOM, method="hybrid", deepSplit = 2, pamRespectsDendro = FALSE, minClusterSize = minModuleSize);
    
    #Get the module labels and the size of each module. Lable 0 is reserved for unassigned genes
    print(table(dynamicMods))
    
    #Plot the module assignment under the dendrogram; note: The grey color is reserved for unassigned genes
    dynamicColors = labels2colors(dynamicMods)
    table(dynamicColors)
    plotDendroAndColors(geneTree, dynamicColors, "Dynamic Tree Cut", dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05, main = "Gene dendrogram and module colors")
    
    #Set the diagonal of the dissimilarity to NA 
    diag(dissTOM) = NA;
    
    #extract modules
    module_colors= setdiff(unique(dynamicColors), "grey")
    modules = lapply(module_colors, function(x){colnames(datExpr)[which(dynamicColors==x)]})
    names(modules) = module_colors
    return(list(dyanmicColors = dynamicColors, modules = modules)) #returns list with module colors, and the modules themselves
  }
  
  # Merge modules with low dissimilarity. Cutoff for dissimilarity merge can be specified
  MergeSimilarModules <- function(datExpr, dynamicColors, geneTree, MEDissThres = 0.5){
    #cacluate eigengenes
    MEList = moduleEigengenes(datExpr, colors=dynamicColors)
    MEs = MEList$eigengenes
    
    # Calculate dissimilarity of module eigengenes
    MEDiss = 1-cor(MEs);
    
    # Cluster module eigengenes
    METree = hclust(as.dist(MEDiss), method = "average");
    
    # Plot the result
    #sizeGrWindow(7, 6)
    plot(METree, main = "Clustering of module eigengenes",
         xlab = "", sub = "")
    abline(h = MEDissThres, lwd=2, col="red")
    
    # Call an automatic merging function
    merge = mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose = 3)
    # The merged module colors
    mergedColors = merge$colors;
    # Eigengenes of the new merged modules:
    mergedMEs = merge$newMEs;
    #plot showing how merged modules exist
    plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
                        c("Dynamic Tree Cut", "Merged dynamic"),
                        dendroLabels = FALSE, hang = 0.03,
                        addGuide = TRUE, guideHang = 0.05)
    
    #extract merged modules
    merged_module_colors= setdiff(unique(mergedColors), "grey")
    merged_modules = lapply(merged_module_colors, function(x){colnames(datExpr)[which(mergedColors==x)]})
    names(merged_modules) = merged_module_colors
    
    return(list(mergedColors = mergedColors, merged_modules = merged_modules)) #returns list with merged colors, and the merged modules themselves
  }
  
  ############################################################################################################
  ## Functions for assaying module significance against background and temporal variability in module score ##
  ############################################################################################################
  
  ## Test to determine if the genes within the module are truly the least dissimilar compared to randomly generated modules of the same size.
  TestModuleSignificance <- function(mod, dissTOM, expr.data, n_perm = 10000, pval = 0.05, n.bin = 10, seed = 42){
    #vectorize the actual distribution of (dis)similarities, and remove zeros!
    set.seed(seed)
    true.diss = as.vector(dissTOM[mod,mod])
    true.diss = true.diss[-which(true.diss == 0)]
    
    #size of module for permutations
    mod.size = length(mod)
    
    #bin all genes by expression
    expr.avg = rowMeans(expr.data)
    expr.avg = expr.avg[order(expr.avg)]
    expr.avg.cut = as.numeric(x = cut2(x = expr.avg, m=round(length(expr.avg)/n.bin)))
    names(expr.avg.cut) = names(expr.avg)
    
    #create a table of binnings of all genes and of our module genes
    all.bin.table = table(expr.avg.cut)
    mod.bin.table = table(expr.avg.cut[mod])
    
    #randomly generate module with same expression binning structure and run t.test and record results
    test.results = data.frame(statistic = rep(NA, n_perm), pval = rep(NA, n_perm)) #container for results
    
    for (i in 1:n_perm){ #by permutation
      random.mod = list() #create an empty list we will fill with gene names (each element will be gene names by bin)
      
      #for each bin of the mod bin table, randomly select that number of genes from the full set with that expression
      for (j in names(mod.bin.table)){ 
        bin.genes = sample(names(expr.avg.cut)[which(expr.avg.cut == as.numeric(j))], mod.bin.table[j], replace = FALSE)
        random.mod[[as.numeric(j)]] = bin.genes #stick those genes into the random.mod list
      }
      #unlist and vectorize the distribution of (dis)similarities (remember to remove zeros)
      random.mod = unlist(random.mod)
      random.diss = as.vector(dissTOM[random.mod,random.mod])
      random.diss = random.diss[-which(random.diss == 0)]
      
      #perform a one-sided wilcox.test and record the statistic and p-value.
      #Note, IMPORTANT: here we perform the test asking if the true diss is LESS THAN the random diss, as we are trying to minimize dissimilarity
      test = wilcox.test(x = true.diss, y = random.diss, alternative = "less")
      test.results[i,] = c(test$statistic, test$p.value)
    }
    
    #correct for multiple hypothesis testing, and then report the proportion of bad tests
    test.results$FDR = p.adjust(test.results$pval, method = "fdr")
    num.failed.tests = sum(test.results$FDR > pval)
    print(paste(paste(num.failed.tests, n_perm, sep="/"), "permutations failed the Mann-Whitney test.", sep=" "))
    
    #is the percentage of failed tests less than or equal to the p-val specified?
    return(num.failed.tests/n_perm <= pval) #returns a vector of booleans indicating if each module was significant based on the specific p.val
  }
  
 
}


#This function just wraps the above with appropriate prompts to identify modules 
#By default this is using the "RNA" assay. Modifying it is trivial but 
#default is set to protect against incorrectly using SCT 
ModID<-function(seuratobj, cluster=NA, prefix="modules", nPCS="", softpower="", combat=FALSE, batch=NA){
  #subset to a cluster or clusters of interest if desired 
  if (sum(!is.na(cluster))>1){test_clus<-subset(seuratobj, idents=cluster)}else{test_clus<-seuratobj}
  #perform feature selection, scaling, and PCA 
  test_clus<-FindVariableFeatures(test_clus, verbose=FALSE)
  test_clus<-ScaleData(test_clus, verbose=FALSE)
  test_clus<-RunPCA(test_clus, verbose=FALSE)
  
  #prompot for number of PCs needed 
  if(nPCS==""){print(ElbowPlot(test_clus))}
  while(nPCS==""){nPCS<-readline("how many NPCs?")}
  nPCS<-as.integer(nPCS)
  #expanding the genes by projecting the dims to more PCs 
  test_clus<-ProjectDim(test_clus)
  
  #gere we'll grab the top 25 features (positive and negative) associated with each dimension 
  genes<-c()
  for (i in 1:nPCS){genelist<-TopFeatures(test_clus, dim = i, nfeatures = 50,balanced = T, projected = T )
  for (i in 1:length(genelist)){genes<-c(genes, genelist[[i]])}}
  #remove duplicates 
  genes<-unique(genes)
  #grab the matrix from the seurat object 
  test_clus<-as.matrix(test_clus@assays$RNA@data[genes,])
  if(combat==TRUE){
    test_clus<-ComBat(test_clus,batch)

  }
  #prompt for softpower 
  FindPower(datExpr=t(test_clus))
  while(softpower==""){softpower<-readline("what softpower?")}
  softpower<-as.integer(softpower)
  #run the remaining functions to actually cut the tree
  test_clus_tom<-ClusterTOM(datExpr = t(test_clus), softPower = softpower)
  test_clus_mods<-CutTOMTree(datExpr = t(test_clus), geneTree = test_clus_tom$geneTree, dissTOM = test_clus_tom$dissTOM, minModuleSize = 10 )
  test_clus_merge_mods<-MergeSimilarModules(datExpr=t(test_clus), dynamicColors = test_clus_mods$dyanmicColors, geneTree = test_clus_tom$geneTree, MEDissThres = 0.5)
  print(test_clus_merge_mods$merged_modules)                                          
  test_clus_merge_mods.isSig = sapply(test_clus_merge_mods$merged_modules, function(module){
    TestModuleSignificance(mod = module, dissTOM = test_clus_tom$dissTOM, expr.data = test_clus,
                           n_perm = 10000, pval = 0.05, n.bin = 10)
  })
  test_clus_merge_mods.isSig = test_clus_merge_mods$merged_modules[test_clus_merge_mods.isSig]
  print(test_clus_merge_mods.isSig)
  names(test_clus_merge_mods.isSig)<-paste(prefix,names(test_clus_merge_mods.isSig ), sep="_")
  return(test_clus_merge_mods.isSig)
}

#lets me wrap a whole group of objects
WrapGroupModID<-function(seuratobjlist, cluster, conds=""){mods<-list()
for (i in 1:length(seuratobjlist)){
  
  print(names(seuratobjlist)[[i]])
  if(length(conds)==1){res<-try(ModID(seuratobjlist[[i]],cluster=cluster, prefix=names(seuratobjlist)[[i]]))}else{
    res<-try(ModID(seuratobjlist[[i]],cluster=cluster, prefix=names(seuratobjlist)[[i]], nPCS =conds[i,1], softpower=conds[i,2]))
  }
  if(inherits(res, "try-error"))
  {
    #error handling code,  just skip this iteration if not modules get found or theres some error in the process
    next
  }
  
  
  mods[[i]]<-res}
return(mods)
}
#this function clusters mods and allows the user to select a cutoff for merging
MergeMods<-function(modslist,height=""){
  #first find the overlap bettween modules 
  modslist<-modslist[!unlist(lapply(modslist, is.null))]
  name<-c()
  for(i in 1:length(modslist)){
    name<-c(name, names(modslist[[i]]))
  }
  modslist<-unlist(modslist, recursive = FALSE)
  overlaps<-list()
  for(i in 1:length(modslist)){
    modoverlap<-list()
    for(j in 1:length(modslist)){
      modoverlap[[j]]<-length(intersect(modslist[[i]], modslist[[j]]))/length(modslist[[i]])

    }
    overlaps[[i]]<-modoverlap
  }
  
  #datawrangle that for plotting
  data<-matrix(unlist(overlaps), nrow = length(overlaps))
  rownames(data)<-colnames(data)<-name
  p1<-heatmap(data, keep.dendro=TRUE)
  print(heatmap(data))
  res2 = as.hclust( p1$Rowv )
  print(plot(res2))
  #ask for user cut 
  if(height==""){goodenough<-FALSE
  while(goodenough==FALSE){height<-readline(prompt ="where to make the cut")
  if(height==""){height=2}
  res<-cutree(res2, h=as.numeric(height))
  print("number of modules is")
  print(max(res))
  print(plot(res2))
  rect.hclust(res2, h=as.numeric(height), border=2:6)
  goodenough<-readline(prompt = "(w)rite mods or (c)ut again?")
  goodenough<-case_when(goodenough=="w"|goodenough=="W" ~TRUE, TRUE ~FALSE)
  }}else{
      res<-cutree(res2, h=as.numeric(height))
      rect.hclust(res2, h=as.numeric(height), border=2:6)
  }
  #merge based on those identified clusters
  mergedmods<-list()
  for( i in 1:max(res)){
    mergedmods[[i]]<-unique(unlist(modslist[res==i]))
  }
  #return modules 
  return(mergedmods)}



#loading data
results<-readRDS( "~/gibbs/DOGMAMORPH/Ranalysis/Objects/20230529CD4ObjAnno.rds")
#setting correct default
DefaultAssay(results)<-"RNA"

```

## module discovery 

The ModID function above is modified to accept a batch variable and to correct with combat. Combat was choosen as opposed to more single cell specific techniques (rPCA, Harmony, FastMNN, etc.) for a few reasons. 
1. Low cell numbers - we don't have enough different cells for our conditions for graph based methods to learn things efficiently. 
2. Need corrected gene values - the above correct in a reduced PCA like space, not the gene expression values themselves. CCA is an alternative here but the concern from point 1 remains. 
3. Reduce chance of overcorrection. Combat's weakness in most applications is that it corrects only weekly and this is not sufficient to overcome tech or batch effects when applying unsupervised methods across those variables. It does not (at least in the studies I've seen) force disparate cell types together in the way that CCA can, which otherwise makes the DE testing have elevated false discovery. 


```{r, modid, fig.width=16, fig.height=9}

results<-subset(results, idents = "Proliferating")
results$batch<-case_when(grepl("4|5|6",results$orig.ident)~1,grepl("7|8|9",results$orig.ident)~2,grepl("10|11|12",results$orig.ident)~3 )
mods<-ModID(results, combat=TRUE, batch=results$batch, nPCS = 8, softpower = 16)
saveRDS(mods, "~/gibbs/DOGMAMORPH/Ranalysis/modules/20230601ProliferatingModules.rds")

for (i in names(mods)){
  write.table(mods[[i]], file=paste0("~/gibbs/DOGMAMORPH/Ranalysis/modules/proliferation/combat_",i,".txt"), quote = FALSE, row.names = FALSE, col.names = FALSE, sep = "")
}

```

## annotating modules 

After correction, we find 4 modules, of those four none appear paritcularly different across treatment, batch, or timepoint. Individual to individual variation is present, but its ultimately not aligned with an interesting variable. 

Further iterations are below either doing the batch effect correction before feature identification or without merging modules based on similarity. In all, it looks like the only two consistent signal (at a module level) in proliferating cells is that:
1. Proliferation
2. Ribosomal/Housekeeping

```{r plotting modules, fig.width=16, fig.height=9}
results<-AddModuleScore(results, mods, name = names(mods))
modnames<-paste0(names(mods), seq(1, length(names(mods))))
FeaturePlot(results, features = modnames, min.cutoff = 'q5', max.cutoff = 'q95')
VlnPlot(results, features = modnames, split.by = "Participant")
VlnPlot(results, features = modnames, split.by = "batch")
VlnPlot(results, features = modnames, split.by = "Treatment")
VlnPlot(results, features = modnames, split.by = "Timepoint")
```



```{r readin enrichr ploting, fig.width=16, fig.height=9}

#borrowing some rushmore colors from the wes anderson color pack
BottleRocket2 = c("#FAD510", "#CB2314", "#273046", "#354823", "#1E1E1E")


#defining a quick function to read in and plot Enrichr 

PlotEnrichment<-function(file, topn=10, returnplot=TRUE, title="Significant Pathways"){
  data<-read.table(file, sep = "\t", header=1)
  data$value<- -log10(data$Adjusted.P.value)
  #order by most signfiicant 
  data$Term<-factor(data$Term, levels = rev(data$Term))
  data<-data[order(data$value, decreasing = TRUE),]
  #subset to top n 
  data<-data[1:topn,]
  plot<-ggplot(data, aes(x=value, y=Term))+geom_bar(stat="identity", fill=BottleRocket2[2])+xlab("-log10(p_adj)")+ylab("Pathway")+ggtitle(title)+theme_classic()
  if(returnplot){return(plot)}
  return(data)
}

todo<-list.files("~/gibbs/DOGMAMORPH/Ranalysis/modules/proliferation/enrichr_combat/")

for (i in todo){
  print(PlotEnrichment(paste0("~/gibbs/DOGMAMORPH/Ranalysis/modules/proliferation/enrichr_combat/",i), topn = 20, title=i))
}

```

# running combat before running module ID 

Rational here is that if we correct for batch before doing variable features and PCA we may be able to discover more true signals. In reality, we end up very consistently recovering proliferation and ribosomal modules only despite different npcs and softpowers. 


```{r combat before PCA}
ModID<-function(seuratobj, cluster=NA, prefix="modules", nPCS="", softpower="", combat=FALSE, batch=NA){
    #subset to a cluster or clusters of interest if desired 
    if (sum(!is.na(cluster))>1){test_clus<-subset(seuratobj, idents=cluster)}else{test_clus<-seuratobj}
    
    #perform feature selection, scaling, and PCA 
    if(combat==TRUE){
        test_clus_2<-test_clus@assays$RNA@data
        test_clus_2<-ComBat(test_clus_2,batch)
        test_clus@assays$RNA@data <- test_clus_2
    }
    test_clus<-FindVariableFeatures(test_clus, verbose=FALSE)
    test_clus<-ScaleData(test_clus, verbose=FALSE)
    test_clus<-RunPCA(test_clus, verbose=FALSE)
    
    #prompot for number of PCs needed 
    if(nPCS==""){print(ElbowPlot(test_clus))}
    while(nPCS==""){nPCS<-readline("how many NPCs?")}
    nPCS<-as.integer(nPCS)
    #expanding the genes by projecting the dims to more PCs 
    test_clus<-ProjectDim(test_clus)
    
    #gere we'll grab the top 25 features (positive and negative) associated with each dimension 
    genes<-c()
    for (i in 1:nPCS){genelist<-TopFeatures(test_clus, dim = i, nfeatures = 50,balanced = T, projected = T )
    for (i in 1:length(genelist)){genes<-c(genes, genelist[[i]])}}
    #remove duplicates 
    genes<-unique(genes)
    #grab the matrix from the seurat object 
    test_clus<-as.matrix(test_clus@assays$RNA@data[genes,])
    
    #prompt for softpower 
    FindPower(datExpr=t(test_clus))
    while(softpower==""){softpower<-readline("what softpower?")}
    softpower<-as.integer(softpower)
    #run the remaining functions to actually cut the tree
    test_clus_tom<-ClusterTOM(datExpr = t(test_clus), softPower = softpower)
    test_clus_mods<-CutTOMTree(datExpr = t(test_clus), geneTree = test_clus_tom$geneTree, dissTOM = test_clus_tom$dissTOM, minModuleSize = 10 )
    test_clus_merge_mods<-MergeSimilarModules(datExpr=t(test_clus), dynamicColors = test_clus_mods$dyanmicColors, geneTree = test_clus_tom$geneTree, MEDissThres = 0.5)
    print(test_clus_merge_mods$merged_modules)                                          
    test_clus_merge_mods.isSig = sapply(test_clus_merge_mods$merged_modules, function(module){
        TestModuleSignificance(mod = module, dissTOM = test_clus_tom$dissTOM, expr.data = test_clus,
                               n_perm = 10000, pval = 0.05, n.bin = 10)
    })
    test_clus_merge_mods.isSig = test_clus_merge_mods$merged_modules[test_clus_merge_mods.isSig]
    print(test_clus_merge_mods.isSig)
    names(test_clus_merge_mods.isSig)<-paste(prefix,names(test_clus_merge_mods.isSig ), sep="_")
    return(test_clus_merge_mods.isSig)
}

#tried npcs/softpower 
#8/18 - Ribosomal and a proliferating 
#7/18 - Ribosomal and a proliferating 
#5/18 - Ribosomal and a proliferating 
#8/10 - Ribosomal and a proliferating 
#8/5 - Ribosomal and a proliferating 
#5/5 - Ribosomal and a proliferating

mods<-ModID(results, combat = TRUE, batch = results$batch, nPCS = 8, softpower = 18)

```

# ID modules without merging similar 

Rational here is that we might be losing interesting variation by merging. Reality was that we ended up with more modules but they mapped back to the same things. 


```{r combat before PCA no merge, fig.width=16, fig.height=9}
ModID<-function(seuratobj, cluster=NA, prefix="modules", nPCS="", softpower="", combat=FALSE, batch=NA){
    #subset to a cluster or clusters of interest if desired 
    if (sum(!is.na(cluster))>1){test_clus<-subset(seuratobj, idents=cluster)}else{test_clus<-seuratobj}
    
    #perform feature selection, scaling, and PCA 
    if(combat==TRUE){
        test_clus_2<-test_clus@assays$RNA@data
        test_clus_2<-ComBat(test_clus_2,batch)
        test_clus@assays$RNA@data <- test_clus_2
    }
    test_clus<-FindVariableFeatures(test_clus, verbose=FALSE)
    test_clus<-ScaleData(test_clus, verbose=FALSE)
    test_clus<-RunPCA(test_clus, verbose=FALSE)
    
    #prompot for number of PCs needed 
    if(nPCS==""){print(ElbowPlot(test_clus))}
    while(nPCS==""){nPCS<-readline("how many NPCs?")}
    nPCS<-as.integer(nPCS)
    #expanding the genes by projecting the dims to more PCs 
    test_clus<-ProjectDim(test_clus)
    
    #gere we'll grab the top 25 features (positive and negative) associated with each dimension 
    genes<-c()
    for (i in 1:nPCS){genelist<-TopFeatures(test_clus, dim = i, nfeatures = 50,balanced = T, projected = T )
    for (i in 1:length(genelist)){genes<-c(genes, genelist[[i]])}}
    #remove duplicates 
    genes<-unique(genes)
    #grab the matrix from the seurat object 
    test_clus<-as.matrix(test_clus@assays$RNA@data[genes,])
    
    #prompt for softpower 
    FindPower(datExpr=t(test_clus))
    while(softpower==""){softpower<-readline("what softpower?")}
    softpower<-as.integer(softpower)
    #run the remaining functions to actually cut the tree
    test_clus_tom<-ClusterTOM(datExpr = t(test_clus), softPower = softpower)
    test_clus_mods<-CutTOMTree(datExpr = t(test_clus), geneTree = test_clus_tom$geneTree, dissTOM = test_clus_tom$dissTOM, minModuleSize = 10 )
    #omiting those that will merge modules so that we can test more individual modules to possibly discover more nuanced differences
    test_clus_merge_mods.isSig = sapply(test_clus_mods$modules, function(module){
        TestModuleSignificance(mod = module, dissTOM = test_clus_tom$dissTOM, expr.data = test_clus,
                               n_perm = 10000, pval = 0.05, n.bin = 10)
    })
    test_clus_merge_mods.isSig = test_clus_mods$modules[test_clus_merge_mods.isSig]
    print(test_clus_merge_mods.isSig)
    names(test_clus_merge_mods.isSig)<-paste(prefix,names(test_clus_merge_mods.isSig ), sep="_")
    return(test_clus_merge_mods.isSig)
}

#tried npcs/softpower 

#7/18 - 3 modules identified - proliferation, ribosomal, and second proliferation 
#8/10 - 5 modules, ribosomal + 4 proliferating 
#5/5 - 3 modules identified - proliferation, ribosomal, and second proliferation 

mods<-ModID(results, combat = TRUE, batch = results$batch, nPCS = 7, softpower = 18)

```

```{r session info}
devtools::session_info()
```

